Updated VCS interface and integration of editor

Janglee123 opened the pull request:
This pr is part of my GSoC project. It has a new UI for commit dock and implemented new features like stage/unstage, discarding changes, branch checkout, fetch/pull/push and etc. 

Testing this branch requires a custom compiled version of [godot-git-plugin](https://github.com/godotengine/godot-git-plugin/pull/37). Both branches are in sync so you can check it out.  

Here is a gif for curiosity. 
<img src="https://user-images.githubusercontent.com/29702428/85693492-ebf17700-b6f3-11ea-9e07-117f7bf7f1e6.gif" width="256">

IronicallySerious commented on editor/editor_vcs_interface.cpp:
```diff
@@ -54,58 +62,107 @@ void EditorVCSInterface::_bind_methods() {
 bool EditorVCSInterface::_initialize(String p_project_root_path) {
 
-	WARN_PRINT("Selected VCS addon does not implement an initialization function. This warning will be suppressed.")
+	WARN_PRINT("Selected VCS addon does not implement  initialization function. This warning will be suppressed.");
```
There are double spaces after `implement`. Minor typo I presume

IronicallySerious replied:
Also I think we are decided on what we are calling the plugin. We have removed places where we refer to it as addon. If you are changing these prompts, might as well change `addon` to `plugin`

IronicallySerious commented on editor/editor_vcs_interface.cpp:
```diff
@@ -54,58 +62,107 @@ void EditorVCSInterface::_bind_methods() {
 
 bool EditorVCSInterface::_is_vcs_initialized() {
-
+	WARN_PRINT("Selected VCS addon does not implement \"" + String(__FUNCTION__) + "\"function. This warning will be suppressed.")
```
Since this exact line repeats quite a lot of times, you can easily put this line in a macro and call that macro wherever you need it

IronicallySerious commented on editor/editor_vcs_interface.cpp:
```diff
@@ -54,58 +62,107 @@ void EditorVCSInterface::_bind_methods() {
 }
 
-bool EditorVCSInterface::_shut_down() {
+bool EditorVCSInterface::_checkout_branch(String branch) {
```
Godot follows generally a convention of `p_` and `r_` (for references) prefix, as you already have been using. There are a few places where this isn't used in this PR. You can find those easily by going through the diff.

This is not high priority but yes, this will be required to get this PR merged in finally

IronicallySerious commented on editor/editor_vcs_interface.cpp:
```diff
@@ -160,6 +223,46 @@ Array EditorVCSInterface::get_file_diff(String p_file_path) {
+}
+
+void EditorVCSInterface::pull() {
+	call("_pull");
```
The `if (is_addon_ready())` check has not be done here and also in a few functions below, unlike the functions above. Is there a specific reason for that?

IronicallySerious commented on editor/editor_vcs_interface.h:
```diff
@@ -51,16 +51,25 @@ class EditorVCSInterface : public Object {
 public:
 	static EditorVCSInterface *get_singleton();
 	static void set_singleton(EditorVCSInterface *p_singleton);
+	int max_commit_fetch = 10;
```
I think this is being used as a constant. You could declare this as const or change this to a macro, so this is not modified by anyone

IronicallySerious commented on editor/icons/icon_refresh.svg:
```diff
@@ -0,0 +1,125 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
The filename convention for icon files seems to be PascalCase. This can be easily changed

Janglee123 replied:
Icon filenames have snake_case for the 3.2 branch. The should be changed while porting it to 4.0.

Janglee123 replied:
@akien-mga  should I compress or remove unnecessary tags from SVG the file.

akien-mga replied:
See https://docs.godotengine.org/en/latest/development/editor/creating_icons.html#icon-optimization

IronicallySerious commented on editor/plugins/version_control_editor_plugin.cpp:
```diff
@@ -133,77 +140,183 @@ void VersionControlEditorPlugin::_initialize_vcs() {
 
 	ERR_FAIL_COND_MSG(!EditorVCSInterface::get_singleton()->initialize(res_dir), "VCS was not initialized");
 
+	EditorVCSInterface::get_singleton()->set_up_credentials(set_up_username->get_text(), set_up_password->get_text());
```
I think we should handle the case if the user does not put anything as username or password. Maybe raise  a warning or a red coloured text if we find empty username and password

IronicallySerious commented on editor/plugins/version_control_editor_plugin.cpp:
```diff
@@ -133,77 +140,183 @@ void VersionControlEditorPlugin::_initialize_vcs() {
 	if (msg == "") {
 
-		commit_status->set_text(TTR("No commit message was provided"));
+		// commit_status->set_text(TTR("No commit message was provided"));
```
You can remove all the code directly if it is not planned to be used

IronicallySerious commented on editor/plugins/version_control_editor_plugin.cpp:
```diff
@@ -133,77 +140,183 @@ void VersionControlEditorPlugin::_initialize_vcs() {
 
-	if (EditorVCSInterface::get_singleton()) {
+	if (!EditorVCSInterface::get_singleton()) {
+		WARN_PRINT("No VCS addon is initialized. Select a Version Control Addon from Project menu");
```
Would really like this renamed from addon to plugin because several users raised questions on what this line actually meant

IronicallySerious commented on editor/plugins/version_control_editor_plugin.cpp:
```diff
@@ -133,77 +140,183 @@ void VersionControlEditorPlugin::_initialize_vcs() {
+	_refresh_commit_list();
 	_refresh_stage_area();
-	_clear_file_diff();
+	// FIX: Editor is not adpoting new changes
```
Minor typo in `adopting` if you are planning to leave this

IronicallySerious commented on editor/plugins/version_control_editor_plugin.cpp:
```diff
@@ -240,42 +353,101 @@ void VersionControlEditorPlugin::_stage_selected() {
+
+		if (scene_extensions.find(file_path.get_extension())) {
+			//FIXIT: Alway opens new tab regardless of scene is open or not.
+			EditorNode::get_singleton()->load_scene(file_path);
```
I am not sure why this is needed. I will build and run and check this again

Janglee123 replied:
The stage and unstage have a button to open the changed file. It opens the same scene again and again in new tabs.

IronicallySerious commented on editor/plugins/version_control_editor_plugin.h:
```diff
@@ -49,46 +51,55 @@ class VersionControlEditorPlugin : public EditorPlugin {
+
+	OptionButton *branch_select;
+
+	ToolButton *fetch_button;
+	ToolButton *pull_button;
+	ToolButton *push_button;
+
+	ToolButton *stage_all_button;
+	ToolButton *unstage_all_button;
+	ToolButton *discard_all_button;
+	ToolButton *refresh_button;
```
What was the rationale behind converting all these to tool buttons?

Janglee123 replied:
Tool buttons are used in other docks and match the editor theme. The normal button has a rectangle box with a border as a background that breaks the theme.

twaritwaikar reviewed (commented):
So far this is awesome! The overall structure is looking well. However you are up for it we could do some internal cleanups like removing the default behavior functions altogether and checking the script directly if it implements a function and then call it, else we do our default procedure in the public functions in EditorVCSInterface. But that is of course not the focus right now.
But overall this looks good :)

Janglee123 committed:
Updated commit dock UI

groud commented on editor/editor_vcs_interface.cpp:
```diff
@@ -45,6 +45,14 @@ void EditorVCSInterface::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("_get_file_diff", "file_path"), &EditorVCSInterface::_get_file_diff);
 	ClassDB::bind_method(D_METHOD("_stage_file", "file_path"), &EditorVCSInterface::_stage_file);
 	ClassDB::bind_method(D_METHOD("_unstage_file", "file_path"), &EditorVCSInterface::_unstage_file);
+	ClassDB::bind_method(D_METHOD("_discard_file", "file_path"), &EditorVCSInterface::_discard_file);
```
I thought we wanted to remove this file completely and move everything in the calling place. Do you want to do this later or abandon the idea ?

IronicallySerious replied:
Actually removing default functions would be the effect of making that shift. The file itself will be needed because it contains the definitions of the public functions that Godot calls into

groud commented on editor/plugins/version_control_editor_plugin.cpp:
```diff
@@ -133,149 +140,278 @@ void VersionControlEditorPlugin::_initialize_vcs() {
+	String msg = commit_message->get_text();
 
-		EditorVCSInterface::get_singleton()->commit(msg);
+	if (msg == "") {
```
I believe it should be possible to commit without message. But indeed we should probably popup a warning.

groud commented on editor/plugins/version_control_editor_plugin.cpp:
```diff
@@ -133,149 +140,278 @@ void VersionControlEditorPlugin::_initialize_vcs() {
-		version_control_dock_button->set_pressed(false);
-	} else {
+	if (!EditorVCSInterface::get_singleton()) {
+		WARN_PRINT("No VCS plugin is initialized. Select a Version Control Addon from Project menu");
```
The two next tests should like use ERR_FAIL_COND_V_MSG instead. This is usually what we do when a function fails.
I think we should consider this as an error (and not a warning) because the commit button should be disabled when the two following conditions are not met.

groud commented on editor/plugins/version_control_editor_plugin.cpp:
```diff
@@ -133,149 +140,278 @@ void VersionControlEditorPlugin::_initialize_vcs() {
+void VersionControlEditorPlugin::_branch_item_selected(int index) {
 
-	if (EditorVCSInterface::get_singleton()) {
+	if (!EditorVCSInterface::get_singleton()) {
```
Same here, this should be an error this it should not be possible to interact with the VCS dock if there is no initialized VCS.

groud commented:
So far the changes look pretty good :)

Janglee123 committed:
improved diff view

Janglee123 commented:
Here is new diff view.  Its showing a diff of the commit with its parent commit. It works same for single file from staged/unstaged changes.  
Split View
![Screenshot from 2020-07-24 14-37-08](https://user-images.githubusercontent.com/29702428/88376743-a9d35800-cdbb-11ea-8ba4-3b080603cc99.png)
Unified View
![Screenshot from 2020-07-24 14-37-51](https://user-images.githubusercontent.com/29702428/88376763-b5268380-cdbb-11ea-9d4e-7be6552a5fca.png)

Calinou commented:
@Janglee123 To get a better-looking line gutter in the VCS diff UI, you could try swapping out the ASCII `|` character for the Unicode character ` ▏`.

Janglee123 commented:
@Calinou I have tried it but RichTextLable is not rendering it.

Janglee123 commented on scene/gui/text_edit.cpp:
```diff
@@ -1255,7 +1270,8 @@ void TextEdit::_notification(int p_what) {
 						if (draw_fold_gutter) {
 							int horizontal_gap = (cache.fold_gutter_width * 30) / 100;
-							int gutter_left = cache.style_normal->get_margin(MARGIN_LEFT) + cache.breakpoint_gutter_width + cache.line_number_w + cache.info_gutter_width;
+							int gutter_left = cache.style_normal->get_margin(MARGIN_LEFT) + cache.breakpoint_gutter_width + cache.line_number_w + cache.info_gutter_width + 2 * cache.gutter_indicator_width;
```
@Paulb23 Could you help me to set margin and space between line no and fold gutters so line change gutter do not overlap on fold gutters?  This is how it looks currently
![Screenshot from 2020-08-07 11-17-30](https://user-images.githubusercontent.com/29702428/89613751-9981a980-d8a0-11ea-9ed4-ea150265a00f.png).
I need 2px space between line no and fold gutters for line change gutters.

Paulb23 replied:
The line number gutter pads by an additional char, the fold gutter currently draws over this padding, so the fold icon draw code needs updating.

However, if possible I would recommend holding on for #40973 first, where this is fixed and in general should make gutters easier.

Janglee123 replied:
@Paulb23 I am currently working the 3.2 branch and I will later port this to master.

IronicallySerious commented on editor/editor_vcs_interface.cpp:
```diff
@@ -54,58 +63,123 @@ void EditorVCSInterface::_bind_methods() {
 bool EditorVCSInterface::_initialize(String p_project_root_path) {
-
-	WARN_PRINT("Selected VCS addon does not implement an initialization function. This warning will be suppressed.")
+	ERR_PRINT("Selected VCS addon does not implement \"" + String(__FUNCTION__) + "\"function. This warning will be suppressed.");
```
Just noticed, `\"function` could be `\" function`

IronicallySerious commented on editor/editor_vcs_interface.h:
```diff
@@ -41,22 +41,48 @@ class EditorVCSInterface : public Object {
 	static EditorVCSInterface *singleton;
 
 	static void _bind_methods();
+	static void _bind_types();
```
This function doesn't seem to be defined anywhere. Consider removing it

IronicallySerious commented on editor/editor_vcs_interface.h:
```diff
@@ -70,14 +96,26 @@ class EditorVCSInterface : public Object {
+	void discard_file(String p_file_path);
 	void commit(String p_msg);
-	Array get_file_diff(String p_file_path);
+	Array get_file_diff(String p_identifier, TreeArea area); // area = 0 unstage, 1 = stage, 2 = commit
```
p_area :)

IronicallySerious commented on editor/editor_vcs_interface.h:
```diff
@@ -70,14 +96,26 @@ class EditorVCSInterface : public Object {
+	void push();
+	void fetch();
+	void set_up_credentials(String p_username, String p_password);
+	Array get_line_diff(String p_file_path, String text);
```
Also p_text

IronicallySerious commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -1736,6 +1738,43 @@ void ScriptTextEditor::_color_changed(const Color &p_color) {
+		script_path = script_path.substr(6);
+	}
+
+	Array line_diff = EditorVCSInterface::get_singleton()->get_line_diff(script_path, code_editor->get_text_edit()->get_text());
+	Map<int, int, TextEdit::GutterIndicator> line_gutters;
+
+	code_editor->get_text_edit()->clear_gutter_indicators();
+	for (int i = 0; i < line_diff.size(); i++) {
+		TextEdit::GutterIndicator indicator_type;
+		Dictionary diff = line_diff[i];
+		int new_start = diff["new_start"];
+		int new_lines = diff["new_lines"];
+		int old_lines = diff["old_lines"];
```
All of this depends on the format of the data received from EditorVCSInterface. Could we arrange all that into a neat little struct? EditorVCSInterface can then handle transferring data from the data received from the plugin and return that to Godot

Janglee123 replied:
Using data structures on both sides would be better IMO.

IronicallySerious replied:
@Janglee123 Yes it would be the best case, but it also needs to be convertible to-and-from a variant to be sent through the Object::call() function

twaritwaikar reviewed (changes requested):
Added a suggestion on the diff data arrangement. Using a dictionary and array of variants is fine to talk to the plugin, but using that data in Godot could be made easier if it was all arranged in a struct, also it will reduce the bugs caused by illegal variant conversions if it were to happen in future contributions.

Janglee123 committed:
Replaced Dictionary/Array with custom structure

IronicallySerious commented on editor/editor_vcs_interface.cpp:
```diff
@@ -237,31 +183,43 @@ void EditorVCSInterface::commit(String p_msg) {
+}
 
+List<EditorVCSInterface::Commit> EditorVCSInterface::get_previous_commits() {
+	List<EditorVCSInterface::Commit> commits = List<EditorVCSInterface::Commit>();
```
The `= List<EditorVCSInterface::Commit>()` can be omitted if you want, although not very important

IronicallySerious commented on editor/editor_vcs_interface.h:
```diff
@@ -56,6 +56,52 @@ class EditorVCSInterface : public Object {
+
+		String old_text;
+		String new_text;
+
+		DiffLine(){};
+	};
+
+	class DiffHunk {
+	public:
+		int new_start;
+		int old_start;
+		int new_lines;
+		int old_lines;
+		List<DiffLine> diff_lines;
+
+		DiffHunk(){};
+	};
+
+	class DiffFile {
+	public:
+		String new_file;
+		String old_file;
+		List<DiffHunk> diff_hunks;
+		DiffFile(){};
+	};
+
+	class Commit {
+	public:
+		String author;
+		String msg;
+		String hex_id;
+		int64_t time;
+	};
+
+	class StatusFile {
+	public:
+		TreeArea area;
+		ChangeType change_type;
+		String file_path;
+	};
```
All these are more suited to be `struct` instead of classes since they are only being used for storage, which means you can remove the `public:` because structs have members public by default. Also you can remove the constructors too because you are not doing anything fancy in them.

But this looks much more well organized than before! Good work!

IronicallySerious commented on editor/editor_vcs_interface.h:
```diff
@@ -92,23 +138,39 @@ class EditorVCSInterface : public Object {
+	DiffFile convert_diff_file(Dictionary p_diff_file);
+	Commit convert_commit(Dictionary p_commit);
+	StatusFile convert_status_file(Dictionary p_status_file);
+
```
All these helper functions can be private if I understand this correctly. I don't think they will need to be called from outside

twaritwaikar reviewed (commented):
This will encourage other contributors to use the EditorVCSInterface better throughout the editor. Good job! I just have a few comments/cleanups you can do.

Janglee123 committed:
Added version control field in project settings
Now, version control editor plugin will auto load plugin on startup once it
initialized by user. Also this opens a window to configure attributes
of plugin and interface. :)

Janglee123 committed:
Added error popup window

Janglee123 committed:
Fixed opened scene not reloading after changes on disk

Janglee123 commented:
I am proposing new changes for the interface. The main reason for these changes is that, before the Version Control System only used by VersionControlEditorPlugin but now it has or will be integrated with other editor plugins like script editor, file system dock etc. 

### Current situation and problems

- `VersionControlEditorPlugin` class manages VC plugins like loading or removing a plugin, and the interface too like below. https://github.com/godotengine/godot/blob/71e1ea771c8141f29b9ab035f8d69d75d4b9f4f8/editor/plugins/version_control_editor_plugin.cpp#L149 This causes the problems described in the last point. 

- `EditorVCSInterface` provides a bridge between editor and plugin. Actually it's not a bridge but the plugin itself as its also the base class for plugins. This is also a cause for the last point. 

- Until `VersionControlEditorPlugin` does not load and set a plugin `EditorVCSInterface::get_singleton()` returns null. It's really odd for a singleton to be null. This gives other editor plugins hard time to work with interface. I have tried to solve this by introducing [vcs_load](https://github.com/godotengine/godot/blob/71e1ea771c8141f29b9ab035f8d69d75d4b9f4f8/editor/plugins/script_text_editor.cpp#L1777) and vcs_unload signals into [EditorNode](https://github.com/godotengine/godot/blob/71e1ea771c8141f29b9ab035f8d69d75d4b9f4f8/editor/editor_node.cpp#L5508) class but it's not a good solution. 

```
+-------------------+        +-----------------------+
|                   +------->+                       |
|  Editor Plugins   |        |  EditorVCSInterface   |
|                   <--------+                       |
+-------------------+        +-----------------------+
```

## Proposed Changes 

- Create a new `EditorVCSPlugin` class for the base class as VCS plugins and make `EditorVCSInterface` singleton always available.
- Move managing plugins and project settings to EditorVCSInterface from VersionControlEditorPlugin. So that VersionControlEditorPlugin only provides GUI hooks for using and configuring EditorVCSInterface. 
```
+-------------------+        +-----------------------+       +--------------------+
|                   +------->+                       +------>+                    |
|  Editor Plugins   |        |  EditorVCSInterface   |       |  EditorVCSPlugin   |
|                   +<-------+                       +<------+                    |
+-------------------+        +-----------------------+       +--------------------+
```
### Pros

- The `EditorVCSInterface::get_singleton()` will not be null and other editor plugins can check the status and connect signals of EditorVCSInterface anytime to work with the interface.
- EditroVCSPlugin can be a subclass of EditorPlugin so that it allows installing VCS plugins from asset store as normal ones.

IronicallySerious commented:
Posting our discussion from matrix here for future reference :)

```
Groud
Well I don't see the point of making the interface always available
If there is no VCS support available,you shouldn't be able to get the VCS
So it's kind of expected the signleton to return Null
you might want, eventually, to create a DummyVCS, but what would this be used for ?

Ironically_Serious
Making it always available would mean you would bake an VCS implementation into the editor, which is not what someone who is not using a VCS would like, or just using a different VCS
Although the implementation wouldn't do anything yeah
I think the EditorVCSInterface should not be treated as a bridge, but rather as a template class (as in the template design pattern and not a C++ template) which delegates the actual work to plugins.
Whether this is the best solution or not, that could be a discussion. But in the comment you are trying to use it as a bridge, like in the mediator pattern
Instead it is more of a template
Also janglee I hope you have completed your evaluation by now? Today is the last day IIRC :)

janglee
so is it fine having have vcs_load and vcs_unload signals in EditorNode? or I should think for another way.
Ironically_Serious yes its mediator pattern, because the `EditorVCSInterface` and `VersionControlEditorPlugin` coupled with each other. I want to make `EditorVCSInterface` independent of `VersionControlEditorPlugin`.  Also I have completed my evaluation, thanks. :)
The new class `EditorVCSPlugin` can be used as template class for plugins.

Groud
janglee: I think it's fine yes

janglee
okay. :)

Groud
A signal is ok :)

Ironically_Serious
With signals yeah it should be fine. It makes EditorVCSInterface fully responsible for VCS stuff and anything else wanting to use it would connect to it
Also I think a past tense signal name would suit better if you are going that route
```

aaronfranke commented:
@Janglee123 Is this still desired? If so, it needs to be [rebased](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase) and tested on the latest 3.2 branch.

ChronicallySerious commented:
@aaronfranke IIRC we had decided a bit later to keep the VCS improvements only for 4.0+, which meant that 3.2 would be limited to the older VCS plugin (i.e. v1.x of [godotengine/godot-git-plugin](https://github.com/godotengine/godot-git-plugin/releases)) and 4.0 would enjoy the new push/pull/fetch and other features implemented by janglee.

This PR was the port of #40129 in case 3.2 would be planned to get the new VCS features as well. While we still wait for the GDNative rewrite for 4.0, I guess this PR can be closed because the decision to port the new VCS features to 3.2 from 4.0 was reverted.

However, I should wait for @groud to confirm

cc @Janglee123

Janglee123 commented:
Yup, I will continue working on 4.0 port once GDnative is available.

Janglee123 committed:
Updated doc

Xrayez commented:
> IIRC we had decided a bit later to keep the VCS improvements only for 4.0+, which meant that 3.2 would be limited to the older VCS plugin (i.e. v1.x of [godotengine/godot-git-plugin](https://github.com/godotengine/godot-git-plugin/releases)) and 4.0 would enjoy the new push/pull/fetch and other features implemented by janglee.

I'm working on Git integration in [Goost](https://goostengine.github.io/), see goostengine/goost#98. I could introduce push/pull functionalities to 3.x as well via module independently ([I also distribute custom Godot builds with Goost](https://goostengine.github.io/download.html)), but the problem is that the features heavily depend on editor functionality as well (since the editor provides a common interface for all VCS providers), so I wouldn't be able to modify the UI to facilitate push/pull features as seen in this PR. Even if there's a way, it would be quite hacky or inconsistent with the proposed PR.

That said, it may be worth to provide dummy implementation for push/pull functionalities in 3.x for v1.0 of Git plugin, both from the VCS interface and UI if core developers fear breaking compatibility with 3.x. Simply disable those push/pull buttons if the v1.0 of Git addon won't provide those features, but nonetheless make it still possible for third-parties to implement push/pull features.

---

Also, @Janglee123 please don't forget to make new proxy/endpoint functions virtual on C++ level as done in #50219 when you rebase 3.x/4.x PRs, so don't changes don't accidentally get lost during rebase, thanks for working on this!

twaritwaikar committed:
Refactor error state logs in EditorVCSInterface

Also renames `addon` to `plugin` in the context of the VCS interface

ChronicallySerious commented on doc/classes/EditorVCSInterface.xml:
```diff
@@ -18,33 +27,20 @@
-				- [code]2[/code]: Renamed
-				- [code]3[/code]: Deleted
-				- [code]4[/code]: Typechanged
+				fetchs changes from remote. For now 'origin' is hardcoded for remote name.
```
`fetchs` -> `Fetches`

ChronicallySerious commented on doc/classes/EditorVCSInterface.xml:
```diff
@@ -84,6 +80,31 @@
+			<return type="void">
+			</return>
+			<description>
+				Pulls changes of corrent branch from remote.
```
`corrent` -> `current`

ChronicallySerious commented on doc/classes/EditorVCSInterface.xml:
```diff
@@ -84,6 +80,31 @@
+			<return type="void">
+			</return>
+			<description>
+				Pushes changes of corrent branch to remote.
```
Same as above ^

ChronicallySerious commented on editor/editor_vcs_interface.cpp:
```diff
@@ -152,6 +255,191 @@ String EditorVCSInterface::get_vcs_name() {
+	ClassDB::bind_method(D_METHOD("is_plugin_ready"), &EditorVCSInterface::is_plugin_ready);
+
+	// API methods that redirect calls to the proxy end points
+	ClassDB::bind_method(D_METHOD("initialize", "project_root_path"), &EditorVCSInterface::initialize);
+	ClassDB::bind_method(D_METHOD("is_vcs_initialized"), &EditorVCSInterface::is_vcs_initialized);
+	// ClassDB::bind_method(D_METHOD("get_modified_files_data"), &EditorVCSInterface::get_modified_files_data);
+	ClassDB::bind_method(D_METHOD("stage_file", "file_path"), &EditorVCSInterface::stage_file);
+	ClassDB::bind_method(D_METHOD("unstage_file", "file_path"), &EditorVCSInterface::unstage_file);
+	ClassDB::bind_method(D_METHOD("discard_file", "file_path"), &EditorVCSInterface::discard_file);
+	ClassDB::bind_method(D_METHOD("commit", "msg"), &EditorVCSInterface::commit);
+	// ClassDB::bind_method(D_METHOD("get_file_diff", "file_path", "area"), &EditorVCSInterface::get_file_diff);
+	ClassDB::bind_method(D_METHOD("shut_down"), &EditorVCSInterface::shut_down);
+	ClassDB::bind_method(D_METHOD("get_project_name"), &EditorVCSInterface::get_project_name);
+	ClassDB::bind_method(D_METHOD("get_vcs_name"), &EditorVCSInterface::get_vcs_name);
+	// ClassDB::bind_method(D_METHOD("get_previous_commits"), &EditorVCSInterface::get_previous_commits);
+	// ClassDB::bind_method(D_METHOD("get_branch_list"), &EditorVCSInterface::get_branch_list);
+	ClassDB::bind_method(D_METHOD("checkout_branch", "branch"), &EditorVCSInterface::checkout_branch);
+	ClassDB::bind_method(D_METHOD("push"), &EditorVCSInterface::push);
+	ClassDB::bind_method(D_METHOD("pull"), &EditorVCSInterface::pull);
+	ClassDB::bind_method(D_METHOD("fetch"), &EditorVCSInterface::fetch);
+	ClassDB::bind_method(D_METHOD("set_up_credentials"), &EditorVCSInterface::set_up_credentials);
+	// ClassDB::bind_method(D_METHOD("get_line_diff"), &EditorVCSInterface::get_line_diff);
```
Are the commented bindings needed?

Janglee123 replied:
Thanks for point it out. Those are need and used. I might have messed while rebasing.

ChronicallySerious commented on editor/plugins/version_control_editor_plugin.cpp:
```diff
@@ -97,219 +110,567 @@ void VersionControlEditorPlugin::popup_vcs_set_up_dialog(const Control *p_gui_ba
+		branch_select->add_item(branch_list[i], i);
+	}
+
+	// TODO: Add new brach feature
+	// branch_select->add_separator();
+	// branch_select->add_item("New Branch");
+	branch_select->select(0); // First branch in the list is current branch
```
Can `branch_select` ever be empty here? If a person makes a branch with `git init --bare` they might have no branches in it.

ChronicallySerious commented on editor/plugins/version_control_editor_plugin.cpp:
```diff
@@ -97,219 +110,567 @@ void VersionControlEditorPlugin::popup_vcs_set_up_dialog(const Control *p_gui_ba
+}
+
+void VersionControlEditorPlugin::_refresh_commit_list() {
+	ERR_FAIL_COND_MSG(!EditorVCSInterface::get_singleton(), "No VCS plugin is initialized. Select a Version Control Addon from Project menu");
```
Addon -> plugin 
:)

ChronicallySerious commented on editor/plugins/version_control_editor_plugin.cpp:
```diff
@@ -97,219 +110,567 @@ void VersionControlEditorPlugin::popup_vcs_set_up_dialog(const Control *p_gui_ba
+}
+
+void VersionControlEditorPlugin::_commit() {
+	ERR_FAIL_COND_MSG(!EditorVCSInterface::get_singleton(), "No VCS plugin is initialized. Select a Version Control Addon from Project menu");
```
There are multiple calls with the same string so you can probably create a function that prints this and call it whenever needed. Probably named `_ensure_vcs_plugin_is_initialized()` or something similar

twaritwaikar reviewed (commented):
Suggested a few refactors that can be done so that the plugin is more consistent with naming and error handling, and pointed a few typos.

Also, there are a couple of merge conflicts so if you can address those as well then I think this would be near to being ready :D 

Would need @akien-mga's advice on if it is suitable for this beta or should we wait for the next 3.x

Xrayez commented on editor/editor_vcs_interface.h:
```diff
@@ -38,45 +38,138 @@
 
 	// Implemented by addons as end points for the proxy functions
 	bool _initialize(String p_project_root_path);
 	bool _is_vcs_initialized();
-	Dictionary _get_modified_files_data();
+	Array _get_modified_files_data();
 	void _stage_file(String p_file_path);
+	void _discard_file(String p_file_path);
 	void _unstage_file(String p_file_path);
 	void _commit(String p_msg);
-	Array _get_file_diff(String p_file_path);
+	Array _get_file_diff(String p_identifier, TreeArea area);
 	bool _shut_down();
 	String _get_project_name();
 	String _get_vcs_name();
+	Array _get_previous_commits();
+	Array _get_branch_list();
+	void _pull();
+	void _push();
+	void _fetch();
+	bool _checkout_branch(String p_branch);
+	void _set_up_credentials(String p_username, String p_password);
```
Should be made `virtual`, see #50219.

Janglee123 committed:
Hide API methods from gdscript

groud reviewed (approved):
A lot of good changes! 
I could not spot any deal-breaking mistake, so approved. :)

akien-mga commented on editor/plugins/version_control_editor_plugin.cpp:
```diff
@@ -355,174 +714,290 @@ const String VersionControlEditorPlugin::get_vcs_name() const {
-	diff_refresh_button->connect("pressed", this, "_refresh_file_diff");
-	diff_hbc->add_child(diff_refresh_button);
+	diff_view_type_select = memnew(OptionButton);
+	diff_view_type_select->add_item("Split", DIFF_VIEW_TYPE_SPLIT);
+	diff_view_type_select->add_item("Unified", DIFF_VIEW_TYPE_UNIFIED);
```
```suggestion
	diff_view_type_select->add_item(TTR("Split"), DIFF_VIEW_TYPE_SPLIT);
	diff_view_type_select->add_item(TTR("Unified"), DIFF_VIEW_TYPE_UNIFIED);
```

akien-mga commented on editor/plugins/version_control_editor_plugin.cpp:
```diff
@@ -355,174 +714,290 @@ const String VersionControlEditorPlugin::get_vcs_name() const {
+
+	GLOBAL_DEF("version_control/autoload_on_startup", false);
+	GLOBAL_DEF("version_control/plugin_name", "");
+	ProjectSettings::get_singleton()->set_custom_property_info("version_control/plugin_name", PropertyInfo(Variant::STRING, "version_control/plugin_name"));
+	ProjectSettings::get_singleton()->set_custom_property_info("version_control/autoload_on_startup", PropertyInfo(Variant::BOOL, "version_control/autoload_on_startup"));
```
These shouldn't be necessary as you're not adding any additional usage hints.

akien-mga commented on editor/editor_vcs_interface.cpp:
```diff
@@ -152,6 +260,191 @@ String EditorVCSInterface::get_vcs_name() {
+}
+
+void EditorVCSInterface::_not_implemented_function(String p_function) {
+	ERR_PRINT("Selected VCS plugin does not implement " + p_function + " function");
```
```suggestion
	ERR_PRINT(vformat("The selected VCS plugin does not implement the \"%s\" function.", p_function));
```

akien-mga commented on editor/editor_vcs_interface.cpp:
```diff
@@ -152,6 +260,191 @@ String EditorVCSInterface::get_vcs_name() {
+
+	ClassDB::bind_method(D_METHOD("is_plugin_ready"), &EditorVCSInterface::is_plugin_ready);
+
+	// // API methods that redirect calls to the proxy end points
+	// ClassDB::bind_method(D_METHOD("initialize", "project_root_path"), &EditorVCSInterface::initialize);
+	// ClassDB::bind_method(D_METHOD("is_vcs_initialized"), &EditorVCSInterface::is_vcs_initialized);
+	// ClassDB::bind_method(D_METHOD("get_modified_files_data"), &EditorVCSInterface::get_modified_files_data);
+	// ClassDB::bind_method(D_METHOD("stage_file", "file_path"), &EditorVCSInterface::stage_file);
+	// ClassDB::bind_method(D_METHOD("unstage_file", "file_path"), &EditorVCSInterface::unstage_file);
+	// ClassDB::bind_method(D_METHOD("discard_file", "file_path"), &EditorVCSInterface::discard_file);
+	// ClassDB::bind_method(D_METHOD("commit", "msg"), &EditorVCSInterface::commit);
+	// ClassDB::bind_method(D_METHOD("get_file_diff", "file_path", "area"), &EditorVCSInterface::get_file_diff);
+	// ClassDB::bind_method(D_METHOD("shut_down"), &EditorVCSInterface::shut_down);
+	// ClassDB::bind_method(D_METHOD("get_project_name"), &EditorVCSInterface::get_project_name);
+	// ClassDB::bind_method(D_METHOD("get_vcs_name"), &EditorVCSInterface::get_vcs_name);
+	// ClassDB::bind_method(D_METHOD("get_previous_commits"), &EditorVCSInterface::get_previous_commits);
+	// ClassDB::bind_method(D_METHOD("get_branch_list"), &EditorVCSInterface::get_branch_list);
+	// ClassDB::bind_method(D_METHOD("checkout_branch", "branch"), &EditorVCSInterface::checkout_branch);
+	// ClassDB::bind_method(D_METHOD("push"), &EditorVCSInterface::push);
+	// ClassDB::bind_method(D_METHOD("pull"), &EditorVCSInterface::pull);
+	// ClassDB::bind_method(D_METHOD("fetch"), &EditorVCSInterface::fetch);
+	// ClassDB::bind_method(D_METHOD("set_up_credentials"), &EditorVCSInterface::set_up_credentials);
+	// ClassDB::bind_method(D_METHOD("get_line_diff"), &EditorVCSInterface::get_line_diff);
```
Dead code?

Janglee123 replied:
We Will allow other plugins to use VCS interface, but not for now.

ChronicallySerious replied:
On second thought, I don't think there is any harm in allowing editor plugins or tool scripts to interact with EditorVCSInterface, if it's just the matter of binding a handful of functions

ChronicallySerious replied:
^ Reminder, these need to be uncommented or deleted

akien-mga commented on editor/editor_vcs_interface.cpp:
```diff
@@ -152,6 +260,191 @@ String EditorVCSInterface::get_vcs_name() {
+
+void EditorVCSInterface::_bind_methods() {
+	// Proxy end points that act as fallbacks to unavailability of a function in the VCS addon
+	ClassDB::bind_method(D_METHOD("_initialize", "project_root_path"), &EditorVCSInterface::_initialize);
+	ClassDB::bind_method(D_METHOD("_is_vcs_initialized"), &EditorVCSInterface::_is_vcs_initialized);
+	ClassDB::bind_method(D_METHOD("_get_vcs_name"), &EditorVCSInterface::_get_vcs_name);
+	ClassDB::bind_method(D_METHOD("_shut_down"), &EditorVCSInterface::_shut_down);
+	ClassDB::bind_method(D_METHOD("_get_project_name"), &EditorVCSInterface::_get_project_name);
+	ClassDB::bind_method(D_METHOD("_get_modified_files_data"), &EditorVCSInterface::_get_modified_files_data);
+	ClassDB::bind_method(D_METHOD("_commit", "msg"), &EditorVCSInterface::_commit);
+	ClassDB::bind_method(D_METHOD("_get_file_diff", "file_path", "area"), &EditorVCSInterface::_get_file_diff);
+	ClassDB::bind_method(D_METHOD("_stage_file", "file_path"), &EditorVCSInterface::_stage_file);
+	ClassDB::bind_method(D_METHOD("_unstage_file", "file_path"), &EditorVCSInterface::_unstage_file);
+	ClassDB::bind_method(D_METHOD("_discard_file", "file_path"), &EditorVCSInterface::_discard_file);
+	ClassDB::bind_method(D_METHOD("_get_previous_commits"), &EditorVCSInterface::_get_previous_commits);
+	ClassDB::bind_method(D_METHOD("_get_branch_list"), &EditorVCSInterface::_get_branch_list);
+	ClassDB::bind_method(D_METHOD("_checkout_branch", "branch"), &EditorVCSInterface::_checkout_branch);
+	ClassDB::bind_method(D_METHOD("_push"), &EditorVCSInterface::_push);
+	ClassDB::bind_method(D_METHOD("_pull"), &EditorVCSInterface::_pull);
+	ClassDB::bind_method(D_METHOD("_fetch"), &EditorVCSInterface::_fetch);
+	ClassDB::bind_method(D_METHOD("_set_up_credentials"), &EditorVCSInterface::_set_up_credentials);
+	ClassDB::bind_method(D_METHOD("_get_line_diff", "file_path", "text"), &EditorVCSInterface::_get_line_diff);
```
Shouldn't these be bound as virtual methods so that they appear in the docs and can be overridden from script?

Janglee123 committed:
Updated bindings

ChronicallySerious commented on editor/plugins/version_control_editor_plugin.cpp:
```diff
@@ -34,27 +34,43 @@
+#include "editor/filesystem_dock.h"
+
+#define CHECK_ADDON_INITIALIZED() \
+	ERR_FAIL_COND_MSG(!EditorVCSInterface::get_singleton(), "No VCS plugin is initialized. Select a Version Control Addon from Project menu.");
```
It would be nice to turn all references to the godot-git-plugin e.g. from `ADDON` to `PLUGIN`, just so that it remains consistent throughout the editor and the plugin.

ChronicallySerious replied:
After the macro name change, you will require a change to the error message here as well: "Select a Version Control addon" -> "Select a Version Control System plugin"

ChronicallySerious commented on doc/classes/EditorVCSInterface.xml:
```diff
@@ -4,112 +4,301 @@
 	</brief_description>
 	<description>
-		Used by the editor to display VCS extracted information in the editor. The implementation of this API is included in VCS addons, which are essentially GDNative plugins that need to be put into the project folder. These VCS addons are scripts which are attached (on demand) to the object instance of [code]EditorVCSInterface[/code]. All the functions listed below, instead of performing the task th ...
+		Used by the editor to display VCS extracted information in the editor. The implementation of this API is included in VCS addons, which are essentially GDNative plugins that need to be put into the project folder. These VCS addons are scripts which are attached (on demand) to the object instance of [code]EditorVCSInterface[/code]. All the functions listed below, instead of performing the task th ...
```
Can also change addons -> plugins here as well

ChronicallySerious commented on doc/classes/EditorVCSInterface.xml:
```diff
@@ -4,112 +4,301 @@
+			<argument index="0" name="branch" type="String">
+			</argument>
+			<description>
+				This method is called when user clicked to switch a branch to [code]branch[/code].
```
This can be present tense here as well, like the other docstrings on this page

ChronicallySerious commented on doc/classes/EditorVCSInterface.xml:
```diff
@@ -4,112 +4,301 @@
 			</return>
 			<description>
-				Returns the name of the VCS if the VCS has been initialized, else return an empty string.
+				Virtual function to get name of VCS plugin.
```
This function returns "Git" IIRC, and in that case it is actually returning the current VCS in use

ChronicallySerious commented on doc/classes/EditorVCSInterface.xml:
```diff
@@ -4,112 +4,301 @@
+			<argument index="1" name="area" type="int">
+			</argument>
+			<description>
+				Used to show file changes in bottom dock. The [code]identifier[/code] can be commit hash or file path based on value of area (see [enum TreeArea]). Returns Array of Dictionary (see [method create_diff_file] and [method add_diff_hunks_into_diff_file]).
```
```suggestion
				Returns an Array of Dictionary objects (see [method create_diff_file] and [method add_diff_hunks_into_diff_file]) containing the changes in a file or a commit hash. The [code]identifier[/code] can be commit hash or file path based on value of area (see [enum TreeArea]).
```

ChronicallySerious commented on doc/classes/EditorVCSInterface.xml:
```diff
@@ -4,112 +4,301 @@
-				- [code]"line_count"[/code] to store an integer containing the number of lines in the line content.
-				- [code]"old_line_number"[/code] to store an integer containing the old line number of the line content.
-				- [code]"offset"[/code] to store the offset of the line change since the first contextual line content.
+				Used to show line change gutters in script editor. [b]Currently not used by editor.[/b] Returns Array of Dictionary (see [method create_diff_hunk]).
```
```suggestion
				Returns an Array of Dictionary objects, each containing a line diff. (see [method create_diff_hunk])
```
I can't seem to recall what the `p_text` argument is for so feel free to add some words about it to this docstring

ChronicallySerious commented on doc/classes/EditorVCSInterface.xml:
```diff
@@ -4,112 +4,301 @@
+			<return type="Array">
+			</return>
+			<description>
+				This method is called when editor wants status of files. Equivalent to [code]git status[/code]. Returns Array of Dictionary (see [method create_status_file]).
```
```suggestion
				Returns Array of Dictionary (see [method create_status_file]) objects, each containing the status data of a modified file. Equivalent to running plain [code]git status[/code].
```

ChronicallySerious commented on doc/classes/EditorVCSInterface.xml:
```diff
@@ -4,112 +4,301 @@
-				- [code]2[/code]: Renamed
-				- [code]3[/code]: Deleted
-				- [code]4[/code]: Typechanged
+				This method is called when editor needs list of previous commits. Returns Array of Dictionary (see [method create_commit]).
```
```suggestion
				Returns Array of Dictionary (see [method create_commit]) objects, each containing the data for a past commit.
```

ChronicallySerious commented on doc/classes/EditorVCSInterface.xml:
```diff
@@ -4,112 +4,301 @@
 			</argument>
 			<description>
-				Initializes the VCS addon if not already. Uses the argument value as the path to the working directory of the project. Creates the initial commit if required. Returns [code]true[/code] if no failure occurs, else returns [code]false[/code].
+				Called when initializing the plugin by editor. Returns [code]true[/code] if successfully initialized or [code]false[/code].
```
```suggestion
				Called when initializing the VCS plugin from the editor. Returns whether or not the plugin was successfully initialized.
```

ChronicallySerious commented on doc/classes/EditorVCSInterface.xml:
```diff
@@ -4,112 +4,301 @@
 			</return>
 			<description>
-				Returns [code]true[/code] if the addon is ready to respond to function calls, else returns [code]false[/code].
+				Called to check if plugin is initialized or not.
```
```suggestion
				Returns whether the VCS plugin has been initialized.
```

ChronicallySerious commented on doc/classes/EditorVCSInterface.xml:
```diff
@@ -4,112 +4,301 @@
+			<return type="void">
+			</return>
+			<description>
+				This method called when user wants to perform a pull operation.
```
```suggestion
				Pulls changes from the remote.
```

ChronicallySerious commented on doc/classes/EditorVCSInterface.xml:
```diff
@@ -4,112 +4,301 @@
 			</return>
 			<description>
-				Returns [code]true[/code] if the VCS addon has been initialized, else returns [code]false[/code].
+				This method called when user wants to perform a push operation.
```
```suggestion
				Pushes changes to the remote.
```

ChronicallySerious commented on doc/classes/EditorVCSInterface.xml:
```diff
@@ -4,112 +4,301 @@
+			<argument index="1" name="password" type="String">
+			</argument>
+			<description>
+				This method called after initializing plugin. Save credentials to future use to perform remote operations.
```
```suggestion
				Called after initializing the plugin. The plugin should save the sent account credentials for future use while performing remote operations.
```

ChronicallySerious commented on doc/classes/EditorVCSInterface.xml:
```diff
@@ -4,112 +4,301 @@
 			</return>
 			<description>
-				Shuts down the VCS addon to allow cleanup code to run on call. Returns [code]true[/code] is no failure occurs, else returns [code]false[/code].
+				Called when user either close the the editor or shut down the VCS plugin.
```
```suggestion
				Shuts down VCS plugin instance. Called when the user either closes the editor or shuts down the VCS plugin through the editor UI.
```

ChronicallySerious commented on doc/classes/EditorVCSInterface.xml:
```diff
@@ -4,112 +4,301 @@
 			</argument>
 			<description>
-				Stages the file which should be committed when [method EditorVCSInterface.commit] is called. Argument should contain the absolute path.
+				This method called when user wants to perform stage operation for [code]file_path[/code].
```
```suggestion
				Stage file present at [code]file_path[/code] to the staged area.
```

ChronicallySerious commented on doc/classes/EditorVCSInterface.xml:
```diff
@@ -4,112 +4,301 @@
 			</argument>
 			<description>
-				Unstages the file which was staged previously to be committed, so that it is no longer committed when [method EditorVCSInterface.commit] is called. Argument should contain the absolute path.
+				This method called when user wants to perform unstage operation for [code]file_path[/code].
```
```suggestion
				Unstage file present at [code]file_path[/code] from the staged area to the unstaged area.
```

ChronicallySerious commented on doc/classes/EditorVCSInterface.xml:
```diff
@@ -4,112 +4,301 @@
+			<argument index="0" name="p_msg" type="String">
+			</argument>
+			<description>
+				Use this to show error message while performing a VCS operation.
```
```suggestion
				Show an error message while performing a VCS operation.
```

ChronicallySerious commented on doc/classes/EditorVCSInterface.xml:
```diff
@@ -4,112 +4,301 @@
+	<signals>
+		<signal name="stage_area_refreshed">
+			<description>
+				Emitted are stage area refreshed by plugin.
```
```suggestion
				Emitted when the stage area is refreshed by the editor.
```
Is changing `plugin` -> `editor` accurate here?

ChronicallySerious commented on editor/editor_vcs_interface.cpp:
```diff
@@ -152,6 +260,190 @@ String EditorVCSInterface::get_vcs_name() {
+	call("_set_up_credentials", p_username, p_password);
+}
+
+Dictionary EditorVCSInterface::create_diff_line(int new_line_no, int old_line_no, String content, String status) {
```
A small number of functions below this are not using the `p_argument_name` convention. Just go over the PR diff once to see if there are any others as well. I don't feel this should affect the operation of the plugin but it wouldn't hurt to check.

twaritwaikar reviewed (changes requested):
Went over the documentation once, suggested some rewords, see if they make things clear or feel free to change the suggestions!

Overall this was very good work!

ChronicallySerious commented on doc/classes/EditorVCSInterface.xml:
```diff
@@ -4,112 +4,301 @@
+			<argument index="0" name="branch" type="String">
+			</argument>
+			<description>
+				called to checkout a [code]branch[/code] in the VCS.
```
```suggestion
				Called to checkout a [code]branch[/code] in the VCS.
```
Just a small documentation error I found while looking again.

ChronicallySerious commented on editor/plugins/version_control_editor_plugin.cpp:
```diff
@@ -97,219 +113,566 @@ void VersionControlEditorPlugin::popup_vcs_set_up_dialog(const Control *p_gui_ba
-		version_control_dock_button->set_pressed(false);
-	} else {
-		WARN_PRINT("No VCS addon is initialized. Select a Version Control Addon from Project menu");
+	// TODO: Add new brach feature.
```
```suggestion
	// TODO: Add new branch feature.
```

ChronicallySerious commented on editor/plugins/version_control_editor_plugin.cpp:
```diff
@@ -97,219 +113,566 @@ void VersionControlEditorPlugin::popup_vcs_set_up_dialog(const Control *p_gui_ba
+	// TODO: Add new brach feature.
+	// branch_select->add_separator();
+	// branch_select->add_item("New Branch");
+	branch_select->select(0); // First branch in the list is current branch
```
We should ideally also check to see if it is a bare repository with 0 branches, but it is really rare to see a user create a bare repository because most Git repositories need to sync to a remote Git repo to actually function as a reliable VCS solution.

twaritwaikar reviewed (approved):
Very excited to see this in action! Since Akien left some comments in his initial review, should probably wait to merge this one yet. But based on my testing, functionally, this PR works great!

akien-mga commented:
Superseded by #53371.
